roadmap/implementers-guide/src/runtime/router.md

```diff
@@ -27,9 +110,108 @@ No initialization routine runs for this module.
 
 ## Routines
 
+The following routines are intended to be invoked by paras' upward messages.
+
+* `init_open_channel(recipient)`:
+  1. Check that the origin of this message is a para. We say that the origin of this message is the `sender`.
+  1. Check that the `sender` is not `recipient`.
+  1. Check that the `recipient` para exists.
+  1. Check that there is no existing open channel request from sender to recipient.
+  1. Check that the sum of the number of already opened HRMP channels by the `sender` (the size of the set found `HrmpEgressChannelsIndex` for `sender`) and the number of open requests by the `sender` (the value from `HrmpOpenChannelRequestCount` for `sender`) doesn't exceed the limit of channels (`config.hrmp_max_parachain_outbound_channels` or `config.hrmp_max_parathread_outbound_channels`) minus 1.
+  1. Reserve the deposit for the `sender` according to `config.hrmp_sender_deposit`
+  1. Add a new entry to `HrmpOpenChannelRequests` and increase `HrmpOpenChannelRequestCount` by 1 for the `sender`.
+      1. Set `sender_deposit` to `config.hrmp_sender_deposit`
+      1. Set `limit_used_places` to `config.hrmp_channel_max_places`
+      1. Set `limit_limit_used_bytes` to `config.hrmp_channel_max_size`
+
+* `accept_open_channel(i)`, `i` - is the index of open channel request:
```

Comment:
I see, so these are all `UpwardMessages`. cool. Noting that implementation depends on https://github.com/paritytech/polkadot/issues/1510

Reply:
Heh, well, that's a subject for discussion. Remember I asked you about introducing special kinds of `UpwardMessage`s instead of relying on dispatching them? That was specifically for this case. I was wondering if it makes sense to add dedicated variants for these messages so that we could reject candidates that fail any of those checks. I didn't go with that way since I thought it might have been too strict. But what do you think?

Reply:
it seems reasonable to reject candidates that fail any checks, actually.

Reply:
Ok, in that case we should relax some requirements to avoid unnecessary rejecting candidates. I.e. [closing](https://github.com/paritytech/polkadot/pull/1503#pullrequestreview-459577136) should not fail if already requested by the other party.

Reply:
but that property might not be easily preserved for e.g. contextual execution.

Reply:
Why not? `close_channel(C)` should be idempotent for the same C even though if it was sent by different origins (At least, as long as we don't allow contextual execution across sessions)